### PR TITLE
expression: MySQL compatible current_user function

### LINKED ISSUE #7613: current_user() function returns incorrect results
Please answer these questions before submitting your issue. Thanks!

1. What did you do?

In TiDB, the `current_user()` function appears to just be an alias of the `user()` function.  In MySQL these have differing behaviors.

2. What did you expect to see?

```
mysql [localhost] {root} ((none)) > CREATE USER 'testuser';
Query OK, 0 rows affected (0.00 sec)

..

mysql [localhost] {testuser} ((none)) > SELECT USER(), CURRENT_USER();
+--------------------+----------------+
| USER()             | CURRENT_USER() |
+--------------------+----------------+
| testuser@localhost | testuser@%     |
+--------------------+----------------+
1 row in set (0.00 sec)
```

Note that `current_user()` shows the same user you would see in `SHOW GRANTS`.  This is useful for corner cases where multiple users could apply (the MySQL privilege system is weird).

3. What did you see instead?

```
MySQL [(none)]> SELECT USER(), CURRENT_USER();
+--------------------+--------------------+
| USER()             | CURRENT_USER()     |
+--------------------+--------------------+
| testuser@127.0.0.1 | testuser@127.0.0.1 |
+--------------------+--------------------+
1 row in set (0.00 sec)
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
Release Version: v2.1.0-rc.1-13-g1255973
Git Commit Hash: 1255973764f72575d71920b95fc0861d1677c6c5
Git Branch: master
UTC Build Time: 2018-08-28 07:35:57
GoVersion: go version go1.10.2 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by morgo**
### What problem does this PR solve? 

Currently current_user() returns the same as user(), which is incorrect.
Fixes #7613

### What is changed and how it works?

Added support for a 'matcheduser' and 'matchedhost'.  This will also be required for statements like `SHOW GRANTS` (without FOR USER option).  See #7614

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change

Side effects

 - Increased code complexity

Related changes

 - Need to update the documentation

### CODE REVIEW

**Comment by morgo** [comment]
I started looking at fixing `SHOW GRANTS`.  It is relatively straightforward, but one question I have for reviewers is if `SessionVars.User` should contain the matched user instead, and have new attributes for the actual user+host.

This would make `SHOW GRANTS` work without modification.  For example see how `SHOW GRANTS` needs a hack to convert the actual user to the matched user.  I suspect there may be other commands that make this mistake:

```
morgo@ryzen:~/go/src/github.com/pingcap/tidb$ git diff
diff --git a/executor/builder.go b/executor/builder.go
index 1f9ef0e0c..51d9d4c44 100644
--- a/executor/builder.go
+++ b/executor/builder.go
@@ -42,6 +42,7 @@ import (
        "github.com/pingcap/tidb/table"
        "github.com/pingcap/tidb/types"
        "github.com/pingcap/tidb/util/admin"
+       "github.com/pingcap/tidb/util/auth"
        "github.com/pingcap/tidb/util/chunk"
        "github.com/pingcap/tidb/util/ranger"
        "github.com/pingcap/tidb/util/timeutil"
@@ -484,7 +485,10 @@ func (b *executorBuilder) buildShow(v *plannercore.Show) Executor {
                is:           b.is,
        }
        if e.Tp == ast.ShowGrants && e.User == nil {
-               e.User = e.ctx.GetSessionVars().User
+               e.User = &auth.UserIdentity{
+                       Username:        e.ctx.GetSessionVars().User.MatchedUsername,
+                       Hostname:        e.ctx.GetSessionVars().User.MatchedHostname,
+               }
        }
        if len(v.Conditions) == 0 {
                return e
diff --git a/planner/core/planbuilder.go b/planner/core/planbuilder.go
index 33c48c11a..f15890462 100644
--- a/planner/core/planbuilder.go
+++ b/planner/core/planbuilder.go
@@ -1576,7 +1576,15 @@ func buildShowSchema(s *ast.ShowStmt) (schema *expression.Schema) {
        case ast.ShowCreateDatabase:
                names = []string{"Database", "Create Database"}
        case ast.ShowGrants:
-               names = []string{fmt.Sprintf("Grants for %s", s.User)}
+
+               if s.User != nil {
+                       names = []string{fmt.Sprintf("Grants for %s", s.User)}
+               } else {
+                       // Default to the session user
+                       // Can't grab the name from here.
+                       names = []string{fmt.Sprintf("Grants for User")}
+               }
+
        case ast.ShowIndex:
                names = []string{"Table", "Non_unique", "Key_name", "Seq_in_index",
                        "Column_name", "Collation", "Cardinality", "Sub_part", "Packed",
```

I have not tested how much breaks yet by setting `SessionVars.User` to the matched user.  Looking for reviewer feedback :-)  Thx.

**Comment by winoros** [comment]
Package `Privilege`'s maintainer is on vacation. There may not be quick response. :(

**Comment by morgo** [comment]
No problems :-)

I noticed one counterpoint today.  In `show processlist` you want the actual identity rather than the matched identity, so this would create a bug if the meaning of `SessionVars.User` was changed.

https://github.com/pingcap/tidb/blob/master/session/session.go#L715-L718

File: session/session.go [hunk 1]
```diff
@@ -1000,16 +1000,20 @@ func (s *session) Auth(user *auth.UserIdentity, authentication []byte, salt []by
 
 	// Check IP.
 	if pm.ConnectionVerification(user.Username, user.Hostname, authentication, salt) {
+		user.MatchedUsername, user.MatchedHostname = pm.ConnectionMatchIdentity(user.Username, user.Hostname)
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply session/session.go]: Agree!

**Comment by morgo** [comment]
I've made the suggested changes, and renamed matched* to auth*.  It looks cleaner - thank you for the feedback.

Should the authUser and AuthHost also be part of the UserPrivileges struct?

**Comment by tiancaiamao** [comment]
LGTM
/run-all-tests

**Comment by tiancaiamao** [comment]
Blocked by https://github.com/pingcap/tidb/pull/7901 which fixes CI

**Pull request merged**